### ISSUE TITLE
Including the Bayan algorithm in cdlib

### ISSUE THREAD

**Issue by saref** [GiulioRossetti/cdlib#212]
**Is your feature request related to a problem? Please describe.**
Detecting communities through modularity maximization

**Reference.**
In case of an algorithm:
There is no paper yet, but we have some info on the Bayan project website: https://bayanproject.github.io/
We have an existing python implementation of the proposed algorithm.

**Describe the solution you'd like**
Following previous email correspondence with Giulio Rosetti, we would like to ask you to consider including the Bayan algorithm in the cdlib.

**Describe alternatives you've considered**
For the purpose that the Bayan algorithm serves (maximizing modularity), we have only found the following method which is too slow to be used on any non-trivial graph:
https://igraph.org/r/doc/cluster_optimal.html
There are of course many heuristic methods for maximizing modularity, but our results indicate that their solutions are, on average, far from a globally optimal solution (see https://bayanproject.github.io/).

**Additional context**
The Bayan algorithm is developed by Hriday Chheda (@Westbrookfc on GitHub) Mahdi Mostajabdaveh (@mahdims) and myself (@saref). We are available to facilitate the steps for adding the Bayan algorithm to CDlib.

**Comment by GiulioRossetti** [comment]
Dear all,
in this [branch](https://github.com/GiulioRossetti/cdlib/tree/bayan) you can find a preliminar version of the requested integration.

Due to the dependency on the Gourobi solver, I opted to include Bayan as an "optional" feature in CDlib. 
The algorithm will be installed along with the library; however, it will be made available to the end user only after he/she has independently installed and configured gouroby and gourobypy.

I removed the warmstart flag since that option triggered a branch containing a syntactic error. 
Moreover, as you can check, I made minor changes to remove unused variables, partially implemented functions' signatures, and similar stuff.

As soon as you confirm that everything works as expected (and as soon as the regression tests are ok), I'll proceed with the merge in the main branch.

**Comment by saref** [comment]
I am writing on behalf of the Bayan team. Thank you very much, Giulio, for your work on this integration request. 

Your solution for ensuring that the user installs Gurobi and then load Bayan as an optional feature of CDlib would be perfect.

Warmstart flag was redundant code, and we should have removed it. We have tested after all the edits you have made, and everything is working perfectly on our end with the exception of one line: 
line 959: "best_lp.upper_bound," is redundant and should be deleted.

Apart from this one change, Bayan is ready to be merged into the main branch from our end.

We are improving Bayan in an agile way. When the improvements become substantial, probably in a few months, we'd like to open another issue for the updated version of Bayan to be considered for CDlib. We acknowledge that our requests take time from you and your team and will minimize the overhead.

**Comment by GiulioRossetti** [comment]
Bayan (integrated through baynapy) will be included as an optional dependency (to simplify conda installation) within CDlib v0.3.0.
